src/storage/src/hummock/compactor.rs

```diff
@@ -777,19 +780,19 @@ impl Compactor {
 
             if level.level_type == LevelType::Nonoverlapping as i32 {
                 debug_assert!(can_concat(&level.table_infos.iter().collect_vec()));
-                table_iters.push(Box::new(ConcatIterator::new(
+                table_iters.push(HummockIteratorUnion::First(ConcatIterator::new(
                     level.table_infos.clone(),
                     self.context.sstable_store.clone(),
                     read_options.clone(),
-                )) as BoxedForwardHummockIterator);
+                )));
             } else {
                 for table_info in &level.table_infos {
                     let table = self
                         .context
                         .sstable_store
                         .load_table(table_info.id, true, &mut stats)
                         .await?;
-                    table_iters.push(Box::new(SstableIterator::create(
+                    table_iters.push(HummockIteratorUnion::Second(SstableIterator::create(
```

Comment:
We can also use ConcatIterator for one table so that we can create merge iterator as `UnorderedMergeIteratorInner<SstableIterator>` rather than `UnorderedMergeIteratorInner<HummockIteratorUnion>`, which could avoid cost of enum.

Reply:
Sounds good. 

But for this PR I think we should keep it more like a refactor and keep the logic the same as original. 

Maybe we can apply this update in #4034?

Reply:
Wrap a single iterator in Concat may still bring some overhead. Not sure which one is better. This reminds me of my considering to remove the `ReceiverExecutor` and only keep the `MergeExecutor` for streaming. 🤣